File: 2017/12/digital-plumber-2.py
# -*- coding: utf-8 -*-
import sys
import collections


def parse_graph(filename):
    with open(filename) as f:
        data = f.read().strip()

    graph = collections.defaultdict(set)
    for line in data.splitlines():
        node, children = parse_line(line)
        graph[node] |= set(children)
    return graph


def parse_line(line):
    node, children = line.split('<->')
    children = children.split(',')
    return int(node), [int(c) for c in children]


def traverse(graph, start=0):
    unvisited = set(graph)

    groups = []
    while unvisited:
        stack = [unvisited.pop()]
        group = set()
        while stack:
            node = stack.pop()
            group.add(node)
            stack.extend(graph[node] - group)
        groups.append(group)
        unvisited -= group

    return groups


def main(filename):
    graph = parse_graph(filename)
    groups = traverse(graph)
    print(len(groups))


if __name__ == '__main__':
    main(sys.argv[1])

File: 2017/12/digital-plumber-2.go
package main

import "bufio"
import "fmt"
import "io"
import "os"
import "strings"
import "strconv"


type Graph map[int][]int


type node struct {
    Value int
    Next *node
}

type Stack struct {
    Top *node
}

func (s *Stack) Push(value int) {
    node := &node{Value: value, Next: s.Top}
    s.Top = node
}

func (s *Stack) Pop() int {
    node := s.Top
    s.Top = s.Top.Next
    return node.Value
}

func (s *Stack) IsEmpty() bool {
    return s.Top == nil
}


func readGraph(filename string) Graph {
    fp, err := os.Open(filename)
    if err != nil {
        panic(err)
    }

    var graph = make(Graph)
    var reader = bufio.NewReader(fp)
    for {
        line, err := reader.ReadString('\n')
        if err != nil {
            if err == io.EOF {
                break
            }
            panic(err)
        }

        var fields = strings.Fields(line)

        var children []int
        for _, field := range fields[2:] {
            var childName = strings.Trim(field, ",")
            child, err := strconv.Atoi(childName)
            if err != nil {
                panic(err)
            }
            children = append(children, child)
        }

        var nodeName = fields[0]
        node, err := strconv.Atoi(nodeName)
        if err != nil {
            panic(err)
        }

        graph[node] = children
    }
    return graph
}


func findUnvisited(visited []bool) (int, bool) {
    for i, v := range visited {
        if !v {
            return i, true
        }
    }
    return 0, false
}


func traverse(graph Graph) (numGroups int) {
    visited := make([]bool, len(graph))

    for {
        start, found := findUnvisited(visited)
        if !found {
            break
        }
        var stack Stack
        stack.Push(start)
        for !stack.IsEmpty() {
            var node = stack.Pop()
            visited[node] = true
            for _, child := range graph[node] {
                if !visited[child] {
                    stack.Push(child)
                }
            }
        }
        numGroups++
    }
    return
}


func main() {
    var filename = os.Args[1]
    var graph = readGraph(filename)
    var numGroups = traverse(graph)
    fmt.Println(numGroups)
}
